File: gitea_repos_master/isphere-classes/worker-classes/worker_classes/keydb/adapter.py
import json
import logging
from typing import List, Any, Optional

from pydash import find, get

from worker_classes.keydb.builder_xml import KeyDBBuilderXML
from worker_classes.keydb.interfaces import (
    ExtraCastFields,
    FieldOrListField,
    Records,
)


class KeyDBAdapter:
    extra_cast_fields: ExtraCastFields = {
        "coordinates": lambda v, r: json.dumps(
            [{"coords": v, "text": get(r, "address", "Location")}]
        )
    }

    def __init__(self, cast_fields: Optional[ExtraCastFields] = None):
        self.extra_cast_fields = {
            **self.extra_cast_fields,
            **(cast_fields if cast_fields else {}),
        }

    def to_key_db(self, data: List[Any], builder_XML: KeyDBBuilderXML) -> Records:
        self.builder_XML = builder_XML

        output = []

        for row in data:
            row_data = []
            for key, value in row.items():
                fields = self._get_fields(key, value, row)
                if not fields:
                    continue
                if isinstance(fields, list):
                    row_data += fields
                elif fields:
                    row_data.append(fields)
            output.append(row_data)

        return output

    def _get_fields(self, key: str, value: Any, row: List[Any]) -> FieldOrListField:
        if not value:
            return None

        skip_field_mask = ["_"]
        skip_field = find(skip_field_mask, lambda f: key.startswith(f))
        if skip_field:
            return None

        reserved_fields = {"list__": self.__cast_list}
        found_reserved = find(list(reserved_fields.keys()), lambda f: key.startswith(f))
        if found_reserved:
            return reserved_fields[found_reserved](key, value)

        if key not in self.builder_XML.description:
            logging.info(
                f'Field "{key}" is not described in XML Builder class. Value: {value}'
            )
            return None

        if key in self.extra_cast_fields:
            return self.builder_XML.create(key, self.extra_cast_fields[key](value, row))

        return self.builder_XML.create(key, value)

    def __cast_list(self, key, value):
        return [self.builder_XML.create(key.replace("list__", ""), v) for v in value]

File: gitea_repos_master/whatsapp/whatsapp-baileys-api/src/logic/keydb/adapter.ts
import logger from '../logger/winston';
import { KeyDBBuilderXML } from './builder_xml';
import { FieldDescription, Records } from 'keydb';

export class KeyDBAdapter {
  private builder_XML: KeyDBBuilderXML;
  public to_key_db(data: any[], builder_XML: KeyDBBuilderXML) {
    this.builder_XML = builder_XML;

    const output: Records = [];

    for (const row of data) {
      const row_data: FieldDescription[] = [];

      for (const [key, value] of Object.entries(row)) {
        const fields = this._get_fields(key, value);
        if (!fields) {
          continue;
        }
        if (Array.isArray(fields)) {
          row_data.push(...fields);
        } else if (fields) {
          row_data.push(fields);
        }
      }

      output.push(row_data);
    }

    return output;
  }

  private _get_fields(key: string, value: any): FieldDescription | null {
    if (!value) {
      return null;
    }

    const skip_field_mask = ['_'];
    const skip_field = skip_field_mask.find((f) => key.startsWith(f));

    if (skip_field) {
      return null;
    }

    const reserved_fields: any = {
      list__: (key: string, value: any[]) =>
        value.map((v) => this.builder_XML.create(key.replace('list__', ''), v))
    };
    const found_reserved = Object.keys(reserved_fields).find((f) =>
      key.startsWith(f)
    );

    if (found_reserved) {
      return reserved_fields[found_reserved](key, value);
    }

    if (!(key in this.builder_XML.description)) {
      logger.warn(
        `Field "${key}" is not described in XML Builder class. Value: ${value}`
      );
      return null;
    }

    return this.builder_XML.create(key, value);
  }
}
